design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+# Design of core collection operations
+
+This document describes an inital design of the core, general data collections operations - e.g.
+operations that aren't specific to a particular collection.
+
+## Assumptions
+
+* Collection information can change at any time, and previous versions are not necessarily
+  accessible.
+  * They should be accessible, at least, to database admins, and possibly via admin API methods.
+* Collections can theoretically be reverted to prior versions at any time.
+* We expect < 1000 collections over the lifetime of the collections infrastructure.
+* There may be many versions of collections as they are updated and adjusted over time.
+  * Including adding new data products to collections.
+* Data product data structures / UIs / etc for collections will usually, but not always,
+  be applicable to other collections as well.
+
+## Collection operations / versions
+
+* Save a collection version.
+  * Administration only operation.
+  * Once saved, a collection version is immutable and cannot be overwritten.
+    * Trying to save a collection with an existing version will result in an error.
+  * A collection version *is not* immediately active.
+    * At some point we may wish to supply admin endpoints for accessing alternate collection
+      versions.
+  * The user specifies the version, which to the collections API is an opaque string. Any
+    semantics are supplied by the user.
+    * Note that this means semantic versions will not sort in order - the DB and API would need
+      specific schemas / code to do that.
```

Comment:
I don't love this. Would prefer to have some sort of ordering in the versions. Even if that means having a "version string" (collections-creating-user provided) and a "version number". I suppose storing the date a version was created could implicitly store the order, but I'm not sure how that would interact with reverted/rollbacked/pre-activated versions.

Reply:
Actually, if we have "source version" as described below, I don't see why we shouldn't enforce strictly ordered versions

Reply:
The creation date of the version implies an ordering. The problem with ordering based on the version tag is that we then have to enforce a specific format for the version tag, which I think of as basically like a git tag. I'd prefer to allow the users of the service to choose their version format.

The rollback mechanism also makes having a version tag with an enforced order not make a lot of sense, since based on the design the version can jump around in the order at will.

> Actually, if we have "source version" as described below, I don't see why we shouldn't enforce strictly ordered versions
I'm not understanding the logic here.

Reply:
Having an easy way to say "this version is before this other one" without passing dates around // display dates to users is quite a convenience boost client-side.

> I'm not understanding the logic here.

Letting the users set the version string arbitrarily makes sense to me from a perspective of "trying to have the version name match the version of the source data", but if we're going to store the source version anyway (i.e. in "source version"), I don't see what benefit we get from arbitrary version names. It seems like we would get more benefit from some sort of sequential versioning scheme, unless I'm missing something.

semver in our collections could provide admins nice visual/scannable inidcators when working with versions. Totally rough-draft spitballing off that, semver would enable guidelines something like the following.
```plaintext
2.11.3
^ ^  ^
| |  |
| |  Increment when a change which does not effect data content is made (description, name, icon) changes
| Increment when collection data content changes, but not the source data version (e.g. updated ETL for that collection)
Increment when collection data content changes to reflect a new source version
```

Reply:
Per out of band discussions, I'm going to write up a doc presenting options for handling collection versions, run it past @dauglyon, and then have the full team discuss it.

Reply:
Added the doc and a decision, so closing this now